app/src/main/java/org/mozilla/focus/browser/VideoVoiceCommandMediaSession.kt

```diff
@@ -0,0 +1,141 @@
+ * - Add this as a [LifecycleObserver]
+ * - Assign [iWebView]
+ * - Add [dispatchKeyEvent] to KeyEvent handling
+ *
+ * To save time, we don't handle audio through either voice or the remote play/pause button: we don't
+ * explicitly handle playback changes ourselves and we mute play/pause events from being received
+ * by the page (see [dispatchKeyEvent]).
+ *
+ * To save time, this implementation doesn't conform to the [MediaSessionCompat] API but still
+ * functions correctly. Conforming to the API requires knowing about playback state, which is
+ * complicated because we need to sync Java state with DOM state. For the current feature set,
+ * this appears to be unnecessary. "Alexa play/pause" will send either event based on the current
+ * playback state so if we lock the initial playback state to PLAYING [1], we can respond to
+ * `onPlay/Pause` events by playing the video if it's paused or pausing the video if it's playing.
+ *
+ * [1]: If the initial playback state is PAUSED or NONE, a music selection voice conversation
+ * overrides our voice commands.
+ */
+class VideoVoiceCommandMediaSession(activity: Activity) : LifecycleObserver {
+
+    private val mediaSession = MediaSessionCompat(activity, MEDIA_SESSION_TAG)
+
+    var iWebView: IWebView? = null // Expected to be set by caller.
+
+    init {
+        val pb = PlaybackStateCompat.Builder()
+                .setActions(SUPPORTED_ACTIONS)
+
+                // See class javadoc for details on STATE_PLAYING.
+                .setState(STATE_PLAYING, PLAYBACK_POSITION_UNKNOWN, 1.0f)
+                .build()
+        mediaSession.setPlaybackState(pb)
+        mediaSession.setCallback(MediaSessionCallbacks())
+        mediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or
+                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)
+    }
+
+    @OnLifecycleEvent(ON_START)
+    fun onStart() {
+        mediaSession.isActive = true
+    }
+
+    @OnLifecycleEvent(ON_STOP)
+    fun onStop() {
+        // Video playback stops when backgrounded so we can end the session.
+        mediaSession.isActive = false
+    }
+
+    @OnLifecycleEvent(ON_DESTROY)
+    fun onDestroy(lifecycleOwner: LifecycleOwner) {
+        // Fragments can be re-used after onDestroy so we must remove the observer
+        // so we don't have two instances when this object is created again.
+        lifecycleOwner.lifecycle.removeObserver(this)
+
+        mediaSession.release()
+    }
+
+    fun dispatchKeyEvent(event: KeyEvent): Boolean = when (event.keyCode) {
+        // If we do nothing, these keys will be received twice: on key down by the MediaSession API
+        // and on key up by the WebView/web page, causing playback state updates to be undone.
+        // Thus we mute the WebView key up events and handle it all ourselves through MediaSession.
+        KEYCODE_MEDIA_PLAY_PAUSE, KEYCODE_MEDIA_PLAY, KEYCODE_MEDIA_PAUSE -> event.action == KeyEvent.ACTION_UP
```

Comment:
Hmmm tricky - so "voice commands" come through as key events? Do using the buttons work too?

Reply:
I should clarify:
- Voice commands call MediaSession callbacks
- Hardware media buttons are swallowed for key down – they call MediaSession callbacks – but on key up they propagate as standard key events (hence they're handled twice)

I'll clarify in the commend.